### ISSUE TITLE
Effect review

### ISSUE THREAD

**Issue by c0ntrol** [c0ntrol/veejay#42]
Review all FX
- ~~"E: Parameter 'Alpha' value 20 out of range 0 - 1" livido plugin parameter default error~~
- _review all FX_
- Transition Fade to Color [improvement]  [see comment here](https://github.com/c0ntrol/veejay/issues/42#issuecomment-305613544)
- Transition Fade to Color [issue] : bright flash on darker part - reproduce - select color > 2 (green) and low framerate (color clipping missing ?)
- ~~review parameter names~~
- Review all keying FX (isolate, complex, rgb, etc) 
- Review page size and step size (per effect) in Reloaded's FX parameter panel
- FX category names ? Total built-in count is 161, frei0r count is 121 plugins,  livido count is 60 plugins making for a pretty long list
- Histogram function for automatic chroma key 
- ~~introduce flags and set if FX requires alpha or writes alpha~~
- ~~figure out why masktransition behaves strangely when parallelized~~
- ~~add button row to fx-chain in reloaded with icons for alpha-in operators~~ moved to issue #73 
- ~~_invent names instead of values for Mode/Type/Operator parameters:_~~
- ~~_motion flag_~~
- ~~add names for Alpha-IN operator value: 0=ignore, 1=A, 2=A and B, 3=A or B, 4=B~~
- ~~refactor libvje initialization, alpha negate cannot be added~~
- ~~add alpha to FX if usefull (Normal Overlay could skip pixels when alpha=0)~~ separate issue #31 
- ~~set subformat to -1 if filter does require to be processed in 4:4:4 planar or if filter does not process chroma channels at all or if filter can be processed in both 4:2:2 planar and 4:4:4 by just setting the uv_len field correctly~~
- ~~set parallel to 1 if filter can be parallized (typically those that do not use global variables in the process function and that do no not need to read pixels outside their work window)~~
- ~~set extra to 1 if filter is a mixing effect~~
- ~~investigate overhead impact of sharing livido plugin workload over multiple cpus~~
- ~~investigate overhead impact of sharing frei0r plugin workload over multiple cpus (needs external white list of filters that can be parallized)~~
- ~~flag livido and frei0r filters that can be parallized~~
- ~~consolidate chains to simplify use ?~~ separate issue #36 
- ~~fix compile warnings~~

Related: issue #31

**Comment by c0ntrol** [comment]
DONE:

* Chroma Key (RGB)
* Matrix Dithering

**Comment by c0ntrol** [comment]
Please list all broken FX you find in 1 comment (edit it)

**Comment by c0ntrol** [comment]
Suspected broken / partly working:
* complex invert
* complex opacity
* complex saturate
* complex threshold
* grey select
* isolate
* keyselect
* rgbkeysmooth

**Comment by d-j-a-y** [comment]
Problem in FX Anim panel ... __can't anim an fx__ ! 

New issue ?

ALso what is this "None" parameter ?

**Comment by d-j-a-y** [comment]
FX Anim  work near 575c0a37d74a6e3a09 ( but problem with toggle keyframming for this chain entry , button who always go off when changing parameter in this panel)

**Comment by d-j-a-y** [comment]
see https://github.com/c0ntrol/veejay/issues/42#issuecomment-156744171

**Comment by c0ntrol** [comment]
It is now possible to name parameter values

And there are many to name!

Add to vj_effect init method:
<pre>
void vje_build_value_hint_list( vj_value_hint_t **hints, int num, int limit, ... );
vj_value_hint_t **vje_init_value_hint_list(int n_params);
</pre>

Examples:
<pre>vj_effect *average_init(int w, int h)
{
    vj_effect *ve = (vj_effect *) vj_calloc(sizeof(vj_effect));
    ve->num_params = 2;
    ve->defaults = (int *) vj_calloc(sizeof(int) * ve->num_params);     /* default values */
    ve->limits[0] = (int *) vj_calloc(sizeof(int) * ve->num_params);    /* min */
    ve->limits[1] = (int *) vj_calloc(sizeof(int) * ve->num_params);    /* max */
    ve->limits[0][0] = 1;
    ve->limits[1][0] = 1000;
    ve->limits[0][1] = 0;
    ve->limits[1][1] = 1;
    ve->defaults[0] = 1;
    ve->defaults[1] = 0;
    ve->parallel = 0; //@ cannot run in parallel
    ve->description = "Average";
    ve->sub_format = 1; 
    ve->extra_frame = 0;
    ve->has_user = 0;
    ve->param_description = vje_build_param_list( ve->num_params, "Number of frames to average", "Mode");
 

    ve->hints = vje_init_value_hint_list( ve->num_params );

    vje_build_value_hint_list( ve->hints, 1, ve->limits[1][1], "Running Average", "Average" );



     return ve;
}
</pre>

<pre>
	ve->hints = vje_init_value_hint_list( ve->num_params );

	vje_build_value_hint_list( ve->hints, 0, ve->limits[1][0],
		"Additive", "Subtractive","Multiply","Divide","Lighten","Hardlight",
		"Difference","Difference Negate","Exclusive","Base","Freeze",
		"Unfreeze","Relative Add","Relative Subtract","Max select", "Min select",
		"Relative Luma Add", "Relative Luma Subtract", "Min Subselect", "Max Subselect",
		"Add Subselect", "Add Average", "Experimental 1","Experimental 2", "Experimental 3",
		"Multisub", "Softburn", "Inverse Burn", "Dodge", "Distorted Add", "Distorted Subtract","Experimental 4", "Negation Divide");
</pre>

</pre>

**Comment by c0ntrol** [comment]
<pre>
baltantv.c:     ve->param_description = vje_build_param_list( ve->num_params, "Stride", "Mode" );
bgsubtract.c:   ve->param_description = vje_build_param_list( ve->num_params, "Threshold", "Mode", "To Alpha");
binaryoverlays.c:       ve->param_description = vje_build_param_list( ve->num_params, "Mode");
cali.c:    ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Subtract Dark Current Only" );
chameleonblend.c:    ve->param_description = vje_build_param_list( ve->num_params, "Mode (Appear,Dissapear)" );
chromamagick.c: ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Value" );
chromamagickalpha.c:    ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Value" );
chromascratcher.c:      ve->param_description = vje_build_param_list(ve->num_params, "Mode", "Value", "Length", "Pingpong" );
chromium.c:     ve->param_description = vje_build_param_list( ve->num_params, "Mode" );
colorhis.c:     ve->param_description = vje_build_param_list ( ve->num_params, "Mode (R,G,B,All)", "Draw","Intensity", "Strength" );
colorshift.c:   ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Value" );
complexsync.c:    ve->param_description = vje_build_param_list( ve->num_params, "Vertical size", "Mode", "Framespeed" );
constantblend.c:        ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Threshold", "Constant" );
contourextract.c:       ve->param_description = vje_build_param_list( ve->num_params, "Threshold", "Mode", "Show image/contour", "Thinning", "Min weight" );
contrast.c:     ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Luma", "Chroma" );
crosspixel.c:   ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Size" );
diff.c: ve->param_description = vje_build_param_list( ve->num_params, "Threshold", "Mode", "Show mask/image", "Thinning" );
diffimg.c:      ve->param_description = vje_build_param_list( ve->num_params,"Mode", "Min threshold", "Max threshold" );
dither.c:       ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Value" );
dupmagic.c:     ve->param_description = vje_build_param_list( ve->num_params, "Mode" );
emboss.c:       ve->param_description = vje_build_param_list( ve->num_params,"Mode" );
fibdownscale.c: ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Fib" );
flare.c:        ve->param_description = vje_build_param_list( ve->num_params , "Mode", "Opacity", "Radius" );
killchroma.c:   ve->param_description = vje_build_param_list(ve->num_params, "Mode" );
lumablend.c:    ve->param_description = vje_build_param_list(ve->num_params, "Mode", "Threshold A", "Threshold B", "Opacity" );
lumakey.c:      ve->param_description = vje_build_param_list(ve->num_params, "Feather", "Min Threshold","Max Threshold","Distance", "Mode" );
lumakeyalpha.c: ve->param_description = vje_build_param_list(ve->num_params, "Selector Mode", "Opacity" );
lumamagick.c:   ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Opacity A", "Opacity B" );
lumamask.c:     ve->param_description = vje_build_param_list(ve->num_params, "X displacement", "Y displacement", "Mode", "Update Alpha" );
magicoverlays.c:        ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Keep or clear color" );
magicoverlaysalpha.c:   ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Keep or clear chroma" );
magicphotos.c:  ve->param_description = vje_build_param_list( ve->num_params, "Photos", "Waterfall", "Mode" );
magicscratcher.c:       ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Scratch frames", "PingPong");
mask.c: ve->param_description = vje_build_param_list( ve->num_params, "Threshold", "Mode" );
mtracer.c:    ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Length");
neighbours.c:   ve->param_description = vje_build_param_list( ve->num_params, "Brush size", "Smoothness", "Mode (Luma/Chroma)" );
neighbours2.c:  ve->param_description = vje_build_param_list( ve->num_params, "Brush size", "Smoothness", "Mode (Luma/Chroma)" );
neighbours3.c:  ve->param_description = vje_build_param_list( ve->num_params, "Line size", "Smoothness", "Mode (Luma/Chroma)" );
neighbours4.c:  ve->param_description = vje_build_param_list( ve->num_params, "Radius", "Distance from center","Smoothness", "Mode (Luma/Chroma)" );
neighbours5.c:  ve->param_description = vje_build_param_list( ve->num_params, "Stroke size", "Smoothness", "Mode (Luma/Chroma)" );
noiseadd.c:     ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Amplification");
noisepencil.c:  ve->param_description = vje_build_param_list(ve->num_params, "Mode", "Amplification", "Min Threshold", "Max Threshold");
pencilsketch.c: ve->param_description = vje_build_param_list(ve->num_params, "Blend Mode", "Min Threshold", "Max Treshold", "Mask" );
photoplay.c:    ve->param_description = vje_build_param_list(ve->num_params, "Photos", "Waterfall", "Mode");
radioactive.c:  ve->param_description = vje_build_param_list(ve->num_params, "Mode", "Zoom ratio", "Strength", "Difference Threshold" );
rawman.c:       ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Value");
reflection.c:   ve->param_description = vje_build_param_list( ve->num_params, "Value 1", "Value 2", "Mode");
rotozoom.c:     ve->param_description = vje_build_param_list(ve->num_params, "Mode", "Rotate", "Zoom" , "Automatic");
sinoids.c:      ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Sinoids");
slice.c:        ve->param_description = vje_build_param_list( ve->num_params, "Slices", "Mode");
slicer.c:       ve->param_description = vje_build_param_list( ve->num_params, "Width", "Height", "Mode"); 
smear.c:        ve->param_description = vje_build_param_list( ve->num_params, "Mode", "Value" );
split.c:        ve->param_description = vje_build_param_list(ve->num_params, "Mode", "Switch");
videoplay.c:    ve->param_description = vje_build_param_list(ve->num_params, "Photos", "Waterfall", "Mode");
water.c:                        "Wavespeed", "Decay", "Mode", "Threshold (motion)");
zoom.c: ve->param_description = vje_build_param_list( ve->num_params, "Width", "Height", "Factor", "Mode", "Update Alpha" );
</pre>

**Comment by d-j-a-y** [comment]
>It is now possible to name parameter values

nice feature !

**Comment by c0ntrol** [comment]
Yes! If you can, you can help by inventing names for some FX ? Maybe photoplay modes ?

**Comment by c0ntrol** [comment]
list of fX that benefit from -ftree-vectorize gcc flags:

<pre>
ctmf/ctmf.c
effects/alphablend.c
effects/alphaflatten.c
effects/alphanegate.c
effects/baltantv.c
effects/binaryoverlays.c
effects/blob.c
effects/boids.c
effects/cali.c
effects/chromamagick.c
effects/chromium.c
effects/colmorphology.c
effects/coloradjust.c
effects/colorshift.c
effects/common.c
effects/contourextract.c
effects/contrast.c
effects/emboss.c
effects/flare.c
effects/ghost.c
effects/levelcorrection.c
effects/lumamagick.c
effects/magicalphaoverlays.c
effects/magicmirror.c
effects/magicoverlays.c
effects/maskstop.c
effects/masktransition.c
effects/mixtoalpha.c
effects/morphology.c
effects/motionmap.c
effects/negatechannel.c
effects/negation.c
effects/neighbours.c
effects/neighbours2.c
effects/neighbours3.c
effects/neighbours4.c
effects/neighbours5.c
effects/overclock.c
effects/pencilsketch.c
effects/perspective.c
effects/photoplay.c
effects/pixelate.c
effects/radioactive.c
effects/rawman.c
effects/reflection.c
effects/rgbkey.c
effects/softblur.c
effects/timedistort.c
effects/toalpha.c
effects/tracer.c
effects/tripplicity.c
effects/uvcorrect.c
effects/videoplay.c
effects/videowall.c
effects/water.c
effects/waterrippletv.c
transitions/fadecolor.c
transitions/fadecolorrgb.c
transitions/transcarot.c
</pre>

**Comment by d-j-a-y** [comment]
FREEZE : pixelate high values on param1 (near 255)

**Comment by c0ntrol** [comment]
Need exact parameter set for:
* CRASH : __Matte Scratcher__ : (parameter 3 == 0) && (parameter 2 == 0) == CRASH
* BEHAVIOR : __Radial Blur__ ghost white lines on pure black zone

**Comment by d-j-a-y** [comment]
CRASH : Matte Scratcher : 
* param 0 : default value
* param 1 : default value
* set param 2 : 0
* set param 3 : 0

BEHAVIOR : Radial Blur ghost white lines on pure black zone
* default values : [test file - _expire in 60d_](https://framadrop.org/r/o9lVaUX3ul#bXX8+N5HOkZ7WCy2Ze114dAQfwm1c+detJej2a6qGnw=)

**Comment by c0ntrol** [comment]
Matte Scratcher fixed (also fixed problem in Overlay Scratcher and Magic Scratcher)

**Comment by c0ntrol** [comment]
Radial Blur cannot be reproduced... Both with and without --enable-debug 

Can you compile with --enable-debug and provide a backtrace ?

**Comment by d-j-a-y** [comment]
> Radial Blur cannot be reproduced... Both with and without --enable-debug

cpu maybe ?

> Can you compile with --enable-debug and provide a backtrace ?

will try to debug and give an eye to values

**Comment by c0ntrol** [comment]
Sorry, a backtrace won't be helpful (Radial Blur does not crash)

**Comment by d-j-a-y** [comment]
update with cpuinfo

**Comment by c0ntrol** [comment]
By coincidence, I am picking up a Core2 Duo CPU tomorrow ..

**Comment by d-j-a-y** [comment]
```
$:veejay  --benchmark 1280x720

I: Benchmark 1280x720 YUVP 4:2:2 (100 frames)
I: run test 'single-threaded memory copy' (8x) on chunks of 1.76 MB:
I: run done: best score for single-threaded memory copy is 977.2380 ms, worst is 1339.2850 ms, average is 1099.4160 ms
I: run test 'single-threaded slow frame' (8x) on chunks of 1.76 MB:
I: run done: best score for single-threaded slow frame is 1711.2410 ms, worst is 1825.0830 ms, average is 1757.5160 ms
I: Using 2 tasks
I: run test 'multi-threaded slow frame' (8x) on chunks of 1.76 MB:
I: run done: best score for multi-threaded slow frame is 39.8240 ms, worst is 41.7520 ms, average is 41.0410 ms
I: run test 'multi-threaded memory copy' (8x) on chunks of 1.76 MB:
I: run done: best score for multi-threaded memory copy is 18.7490 ms, worst is 19.7070 ms, average is 19.3750 ms
```
nota : i run veejay in 800x600

**Comment by c0ntrol** [comment]
Looks about right to me

**Comment by c0ntrol** [comment]
Did you do a clean build already ? I do not think it is CPU, the Radial Blur effect is not SIMD optimized.

Please find out:
- If you have a stale build
- If it happens on all resolutions
- If it happens with a particular source (image, dummy etc)

You could send me the complete disassembly of libvje/effects/.libs/radialblur.o
<pre>
 objdump -d $FILE
</pre>

**Comment by c0ntrol** [comment]
internal compiler error when compiling on dell poweredge 2650 (dual xeon, 32 bit system) on
* libvje/effects/rgbkey.c (color_distance func, last return statement)

impacts other effects as well

reg_stack.c:689

slackware 14.1 (32 bit) 

todo: verify with different version of gcc compiler or with different 32 bit linux distibution

**Comment by c0ntrol** [comment]
effects/rgbkey.c:277:1 internal compiler error in replace_reg, at reg-stack.c:690

devuan 3.16.7-ckt20-1+dev8u1
gcc: 4.9.2-10

**Comment by d-j-a-y** [comment]
**FREEZE** on pixelate (p0 > ) seems to have come back.

**Comment by c0ntrol** [comment]
confirmed, u_v is 0 so loop takes forever

````
(gdb) info threads
  Id   Target Id         Frame 
  4    Thread 0x7f5b998c3700 (LWP 1304) "veejay" 0x00007f5ba81ae31d in poll ()
   from /lib64/libc.so.6
  3    Thread 0x7f5b98eec700 (LWP 1305) "veejay" 0x00007f5ba8697c44 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  2    Thread 0x7f5b93fff700 (LWP 1306) "veejay" 0x00007f5bb33a006a in pixelate_apply (frame=0x1aa8680, w=720, h=576, vv=512) at effects/pixelate.c:78
* 1    Thread 0x7f5bb3b08b40 (LWP 1303) "veejay" 0x00007f5ba81cc764 in clock_nanosleep () from /lib64/libc.so.6
(gdb) t 2

[Switching to thread 2 (Thread 0x7f5b93fff700 (LWP 1306))]
#0  0x00007f5bb33a006a in pixelate_apply (frame=0x1aa8680, w=720, h=576, 
    vv=512) at effects/pixelate.c:78
78              for (i = 0; i < uv_len; i+=u_v) {
(gdb) print u_v
$1 = 0
 ````

**Comment by d-j-a-y** [comment]
@c0ntrol 

> https://github.com/c0ntrol/veejay/issues/42#issuecomment-168383981 ( internal compiler error in replace_reg, at reg-stack.c:690)

Any news / tips ?

**Comment by nelburg** [comment]
No, seems only to happen on 32 bit systems and also depends on gcc version

**Comment by d-j-a-y** [comment]
> No, seems only to happen on 32 bit systems and also depends on gcc version

maybe some trouble in configure in fixing proc arch ?

**Comment by d-j-a-y** [comment]
YEP found the heck !!! it's from [fast_sqrt](https://github.com/c0ntrol/veejay/blob/master/veejay-current/veejay-server/libvje/effects/common.h#L94) !!!! i have commented the line in rgbkey, and know the problem come at ripple who also use fast_sqrt. `#define AMD64` ?

**Comment by nelburg** [comment]
Yes that is correct 👍 
''''
# define sin_cos(si, co, x)     asm ("fsincos" : "=t" (co), "=u" (si) : "0" (x))
# define fast_sqrt(res__,x) 	asm ("fsqrt" : "=t" (res__) : "0" (x)) 
# define fast_sin(res__,x)	asm ("fsin" : "=t" (res__) : "0" (x))
# define fast_cos(res__,x)	asm ("fcos" : "=t" (res__) : "0" (x))
# define fast_abs(res__,x)	asm ("fabs" : "=t" (res__) : "0" (x))
# define fast_exp(res__,x)	asm ("fexp" : "=t" (res__) : "0" (x))
'''' 

should only be defined for 64 bit system, both intel and amd.
The ifdef wrapping it looks not OK (missing #ifdef for 64 bit) and it should fallback to the simple cos/sqrt/sin versions if 32 bit or anything else non x86

**Comment by nelburg** [comment]
Quick fix (for 32 bit) is to only keep
````
# define sin_cos(si, co, x)     si = sin(x); co = cos(x)
# define fast_sqrt( res,x ) res = sqrt(x) 
# define fast_sin(res,x ) res = sin(x)
# define fast_cos(res,x ) res = cos(x) 
# define fast_abs(res,x ) res = abs(x)
# define fast_exp(res,x ) res = exp(x)
````

**Comment by d-j-a-y** [comment]
Transition Fade to Color [improvement]

* named mode
* color by rgb
* more mode ? (ping pong mode)
* review frame lenght

**Comment by c0ntrol** [comment]
close issue, obsolete, create issue for each effect that does not work properly
